components/prism-solidity.js

```diff
@@ -0,0 +1,37 @@
+Prism.languages.solidity = Prism.languages.extend('javascript', {
+	'keyword': [
+		/\b(?:assert|break|constant|continue|do|else|emit|external|for|function|if|import|inherited|internal|is|memory|new|payable|private|public|pure|require|returns?|revert|selfdestruct|storage|suicide|this|throw|var|view|while|_)\b/,
+		{
+			pattern: /\b(?:contract|interface|library|using|struct|function|modifier)\s+[A-Za-z_]\w*(?:\s+is\s+(?:[A-Za-z_][,\s]*)*)?\b/,
+			inside: {
+				'variable': {
+					pattern: /\b(contract|interface|library|using|struct|function|modifier)\s+[A-Za-z_]\w*(?:\s+is\s+(?:[A-Za-z_][,\s]*)*)?\b/,
+					lookbehind: true
+				}
+			}
+		},
+	],
+	'constant': [
+		{
+			pattern: /\b(?:address|string|bytes\d*|int\d*|uint\d*|bool|u?fixed\d+x\d+)\b(?:\s+(?:indexed\s+)?[A-Za-z_]\w*\s*[,)])?/,
+			inside: {
+				'attr-name' : {
+					pattern: /\b(address|string|bytes\d*|int\d*|uint\d*|bool|u?fixed\d+x\d+)\b(?:\s+(?:indexed\s+)?[A-Za-z_]\w*\s*[,)])?/,
+					lookbehind: true,
+					inside: {
+						punctuation: /[(){};:]/
+					}
+				}
+			}
+		},
+		{
+			pattern: /\bmapping\s*\(.*\s+=>\s+.*\)(?:\s+(?:private|public|internal|external|inherited))?\s+[A-Za-z_]\w*\b/,
```

Comment:
I have a problem with `.*` as the dot also includes most of `\s+=>\s+` which invites catastrophic backtracking.
What exactly is `.*` supposed to match?

Reply:
The first .* is supposed to be a builtin type, the second one may be any type, even another mapping. Don't know if there is a better way to express this.

Reply:
For more information, this is the language specification: https://solidity.readthedocs.io/en/v0.5.7/types.html#mapping-types

Reply:
I don't what the most complex type name could be but wouldn't `\w+` instead of `.*` also work? So `\(\s*\w+\s*=>\s*\w+\s*\)`.

And if we were to allow _any_ content, then `\([^()]*\)` might be what we want.

Reply:
It is possible to have a mapping of mapping, for e.g.: `mapping(address => mapping(uint => customStruct[])) someName`, and so on... how can I model this recursion on prism? Is there some good example I can follow?

Reply:
Matching such expression correctly requires a context-free grammar, so we can't match them with regexes. It's fundamentally impossible.
But we can approximate by matching a limited number of recursion levels.

E.g. `\((?:[^()]|\((?:[^()]|\([^()]*\))*\))*\)` will match recusive bracket expression for up to a maximum depth of 3.

This is how we usually deal with nested expression.